libmscore/changeMap.cpp

```diff
@@ -0,0 +1,487 @@
+                              continue;
+                              }
+                        else {
+                              currentRampStart = event.tick;
+                              currentRampEnd = event.etick;
+                              startsInRamp.push_back(true);
+                              }
+                        }
+                  else {
+                        currentRampStart = event.tick;
+                        currentRampEnd = event.etick;
+                        inRamp = true;
+                        startsInRamp.push_back(false);
+                        }
+
+                  endPoints.push_back(std::make_pair(event.tick, event.etick));
+                  }
+            else if (event.type == ChangeEventType::FIX) {
+                  if (inRamp) {
+                        if (event.tick != currentRampStart && event.tick != currentRampEnd && lastFix != event.tick) {
+                              // delete, this event is enveloped or at the same point as another fix
+                              remove(event.tick, event);
+                              continue;
+                              }
+                        }
+
+                  lastFix = event.tick;
+                  }
+            }
+
+      //
+      // Stage 2: readjust lengths of any colliding ramps:
+      //
+      int j = -1;
+      for (auto& event : values()) {
+            if (event.type != ChangeEventType::RAMP)
+                  continue;
+
+            j++;
+            if (!startsInRamp[j])
+                  continue;
+
+            // Take a copy of the event...
+            auto tempChangeEvent = event;
+            remove(event.tick, event);
+            tempChangeEvent.tick = endPoints[j-1].second;
+            // ...and move it to its new tick position
+            insert(tempChangeEvent.tick, tempChangeEvent);
+            }
+
+      //
+      // Stage 3: cache start and end values for each ramp
+      //
+      for (auto i = begin(); i != end(); i++) {
+            auto& event = i.value();
+            if (event.type != ChangeEventType::RAMP)
+                  continue;
+
+            // Phase 1: cache a start value for the ramp
+            // Try and get a fix at the tick of this ramp
+            bool foundFix = false;
+            for (auto& currentChangeEvent : values(event.tick)) {
+                  if (currentChangeEvent.type == ChangeEventType::FIX) {
+                        event.cachedStart = currentChangeEvent.value;
+                        foundFix = true;
+                        break;
+                        }
+                  }
+
+            // If there isn't a fix, use from the last event:
+            //  - the cached end value if it's a ramp
+            //  - the value if it's a fix
+            if (!foundFix) {
+                  if (i != begin()) {
+                        auto prevChangeEventIter = i;
+                        prevChangeEventIter--;
```

Comment:
As this is a multimap, `prevChangeEventIter--` may result in an iterator pointing to another event at the same tick. Is it what is intended here?

Reply:
Since in this block `foundFix` is false, and there can be at most two events at any one tick (a fix and a ramp), it is certain that there is only one event at this point. So, when we decrement the iterator it will definitely be the event at the last tick position before.

Reply:
Sound good?